### PR TITLE
Add the state machine of updating custom installation id in the local cache and update to Firebase Segmentation backend. 

### PR DESCRIPTION
**Pull request by diwu-arete**
Close review needed for FirebaseSegmentation.java

CL also contains unit tests. 
The http client is not implemented yet.

### CODE REVIEW

**Comment by diwu-arete** [comment]
This is a pretty large PR, but the key part is just FirebaseSegmentation.java which needs a close review.

File: firebase-segmentation/src/main/java/com/google/firebase/segmentation/FirebaseSegmentation.java [hunk 1]
```diff
@@ -55,7 +79,161 @@ public static FirebaseSegmentation getInstance(@NonNull FirebaseApp app) {
     return app.get(FirebaseSegmentation.class);
   }
 
-  Task<Void> setCustomInstallationId(String customInstallationId) {
-    return Tasks.forResult(null);
+  Task<Void> setCustomInstallationId(@Nullable String customInstallationId) {
```
Comment (vkryachko) [inline firebase-segmentation/src/main/java/com/google/firebase/segmentation/FirebaseSegmentation.java]: should the return type be `@NonNull public Task<Void>` instead?

Comment (vkryachko) [reply firebase-segmentation/src/main/java/com/google/firebase/segmentation/FirebaseSegmentation.java]: Also what are the thread-safety/concurrency requirements for this method(including the ones it calls)? I looks like multiple calls to this from different threads could lead to races

File: firebase-segmentation/src/main/java/com/google/firebase/segmentation/FirebaseSegmentation.java [hunk 2]
```diff
@@ -55,7 +79,161 @@ public static FirebaseSegmentation getInstance(@NonNull FirebaseApp app) {
...
+                        CustomInstallationIdCache.CacheStatus.PENDING_UPDATE)));
+
+    // Start requesting backend when first cache update is done.
+    Task<Code> backendRequestResultTask =
+        firstUpdateCacheResultTask.onSuccessTask(
+            firstUpdateCacheResult -> {
+              if (firstUpdateCacheResult) {
+                String iid = instanceIdResultTask.getResult().getId();
+                String iidToken = instanceIdResultTask.getResult().getToken();
+                return backendServiceClient.updateCustomInstallationId(
```
Comment (vkryachko) [inline firebase-segmentation/src/main/java/com/google/firebase/segmentation/FirebaseSegmentation.java]: What if the device is offline, do you want to hold on to the new value and sync when device gets connected or do you want to throw?

File: firebase-segmentation/src/main/java/com/google/firebase/segmentation/FirebaseSegmentation.java (hunk 1 above)
Comment (diwu-arete) [reply firebase-segmentation/src/main/java/com/google/firebase/segmentation/FirebaseSegmentation.java]: 1)  Yes, definitely it should be nonnull public

2)  Change it to be synchronized

File: firebase-segmentation/src/main/java/com/google/firebase/segmentation/FirebaseSegmentation.java (hunk 2 above)
Comment (diwu-arete) [reply firebase-segmentation/src/main/java/com/google/firebase/segmentation/FirebaseSegmentation.java]: I have thought this case.

If the device is offline, I would
1) the backendServiceClient return a error
2) the cache status will remain in PENDING_UPDATE or PENDING_CLEAR
3) have a separate listener that retry the PENDINGs when device is back online.

WDYT?

Comment (vkryachko) [reply firebase-segmentation/src/main/java/com/google/firebase/segmentation/FirebaseSegmentation.java]: Sounds good.

**Pull request merged**